runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+            while (allTasks.isEmpty()) {
+                allTasks.wait();
+            }
+        }
+    }
+
+    private List<StreamingTask> getTasks(StreamStatus status, Predicate<StreamingTask> predicate) {
+        List<StreamingTask> result = new ArrayList<>();
+        for (StreamingTask task : allTasks.values()) {
+            if (task.getStatus() == status && predicate.test(task)) {
+                result.add(task);
+            }
+        }
+        return result;
+    }
+
+    private List<StreamingTask> getTasks(StreamStatus status) {
+        return getTasks(status, t -> true);
+    }
+
+    private List<StreamAddressRange> getPollQueries(List<StreamingTask> tasks) {
+        List<StreamAddressRange> pollRequests = new ArrayList<>(tasks.size());
+        for (StreamingTask task : tasks) {
+            DeltaStream stream = task.getStream();
+            pollRequests.add(new StreamAddressRange(stream.getStreamId(), Address.MAX, stream.getMaxAddressSeen()));
+        }
+        return pollRequests;
+    }
+
+    private void poll(List<StreamingTask> tasks) {
+
+        List<StreamAddressRange> queries = getPollQueries(tasks);
+
+        // In order to increase the efficiency of batching, we need to coalesce all queries before batching, that
+        // way each batch will have unique queries. If we coalesce after batching, then its possible to have common
+        // stream queries across batches, this means that the same stream is queried multiple times, but it could
+        // have only been polled once.
+
+        List<StreamAddressRange> coalesced = coalesce(queries);
+        List<List<StreamAddressRange>> batches = Lists.partition(coalesced, pollBatchSize);
+
+        Map<UUID, StreamAddressSpace> allQueryResults = new HashMap<>();
+
+        for (List<StreamAddressRange> batch : batches) {
+            long pollStartTime = System.nanoTime();
+            Map<UUID, StreamAddressSpace> res = sequencerView.getStreamsAddressSpace(coalesced);
+            MicroMeterUtils.time(Duration.ofNanos(System.nanoTime() - pollStartTime), "StreamPollingScheduler.poll");
+            Preconditions.checkState(res.size() == batch.size());
+            allQueryResults.putAll(res);
+        }
+
+        Preconditions.checkState(tasks.size() == queries.size());
+
+        for (int idx = 0; idx < tasks.size(); idx++) {
+            StreamingTask task = tasks.get(idx);
+            StreamAddressRange taskQuery = queries.get(idx);
+            Preconditions.checkState(task.getStream().getStreamId().equals(taskQuery.getStreamID()));
+            Preconditions.checkState(allQueryResults.containsKey(taskQuery.getStreamID()),
+                    "StreamAddressSpace missing for %s", task.getStream().getStreamId());
+
+            StreamAddressSpace sas = allQueryResults.get(task.getStream().getStreamId()).getAddressesInRange(taskQuery);
+            task.getStream().refresh(sas);
+        }
+    }
+
+    @Data
+    class Interval {
```

Comment:
static?

Reply:
Since it only holds two longs, it doesn't really matter.

Reply:
Changed.